### PR TITLE
Multisig support

### PR DESCRIPTION
**Pull request by gege251**
Allowing contracts to use multiple signers.

For this the user of this library would need to create a new folder called `signing-keys` and put his signing keys in the following format: `signing-key-PUBKEYHASH.skey` so that the pab can find the corresponding keys. 

I used the plutus `UnbalancedTx` and `Tx` types to achieve this functionality, however due to the differences between the cardano-wallet and the cardano-cli, it is not as straightforward as one would think. First, we read private keys in the given folder, and create a `Map PubKeyHash PrivateKey` from it. We use this in the prebalancing step, to sign the transaction (insert signatures to a `Map PubKey Signature`). However, this signature is actually invalid, as the transaction is not fully balanced, and the private key is incorrect (see the reason below). Then, in the build/submit step, we use only the pub keys from the signature map to get the signing key file paths for the cli.

Reading private keys is problematic, because the cardano-wallet (and also the plutus type for PrivateKey) uses a so called extended payment key, which consists of a 64 byte private key, a public key and a chain index, while the payment keys used by the cardano-cli have shorter 32 byte private keys. This means that the signing key files cannot be used in plutus. I used a hacky solution here: filled up the remaining 32 bytes of the private with zeros, and added the correct pub key for the short payment key, so each time I call the `toPublicKey` I get a correct value. This then can be used for mapping between the above mentioned types.

I included a simple test for multisig in the test framework PR (https://github.com/mlabs-haskell/mlabs-pab/pull/6)

Solves: https://github.com/mlabs-haskell/mlabs-pab/issues/4

### CODE REVIEW

**Comment by gege251** [comment]
I included a simple test for multisig in the test framework PR (#6)

File: README.md [hunk 1]
```diff
@@ -87,18 +87,29 @@ To run the fake PAB, you need to prepare a few more things:
...
 
 The fake PAB consists of the following modules:
 
-- **LiquidityBridgeFakePAB** main entry point, Contract re-implementation, handling websocket connections and http endpoints
-- **LiquidityBridgeFakePAB.Contract** handling contract effects by creating the necessary files and calling cardano-cli commands (a few effects are mocked)
-- **LiquidityBridgeFakePAB.CardanoCLI** wrappers for cardano-cli commands
+- **MLabsPAB** main entry point
+- **MLabsPAB.Server** Servant server, handling http endpoint calls and websockets
+- **MLabsPAB.Contract** handling contract effects by creating the necessary files and calling cardano-cli commands (a few effects are mocked)
+- **MLabsPAB.PreBalanc** doing some preparations so the cli can process the rest (non-ada asset balancing, addig tx inputs, adding minimum lovelaces, add signatories)
```
Comment (samuelWilliams99) [inline README.md:106]:
```suggestion
- **MLabsPAB.PreBalance** doing some preparations so the cli can process the rest (non-ada asset balancing, addig tx inputs, adding minimum lovelaces, add signatories)
```

File: nix/ci.nix [hunk 2]
```diff
@@ -1,47 +1,21 @@
...
-, deferPluginErrors ? true
-, doCoverage ? false
-}:
+, plutus ? import sources.plutus { }, deferPluginErrors ? true
+, doCoverage ? false }:
 let
   project = import ./haskell.nix {
     inherit sourcesFile system sources plutus deferPluginErrors doCoverage;
   };
   npmlock2nix = plutus.pkgs.callPackage sources.npmlock2nix { };
```
Comment (samuelWilliams99) [inline nix/ci.nix:9]: Shouldn't need this either

File: src/MLabsPAB/Files.hs [hunk 3]
```diff
@@ -95,6 +120,51 @@ writeAll pabConf policyScripts validatorScripts datums redeemers =
       , mapM (writeRedeemerJsonFile pabConf) redeemers
       ]
 
+readPrivateKeys :: PABConfig -> IO (Either Text (Map PubKeyHash PrivateKey))
+readPrivateKeys pabConf = do
+  files <- listDirectory $ Text.unpack pabConf.pcSigningKeyFileDir
+  let sKeyFiles =
+        map (\filename -> Text.unpack pabConf.pcSigningKeyFileDir ++ "/" ++ filename) $
+          filter (`endsWith` ".skey") files
```
Comment (samuelWilliams99) [inline src/MLabsPAB/Files.hs:128]: Could use `isExtensionOf` here instead, saves a custom def

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]

File: src/MLabsPAB/Files.hs (hunk 3 above)
Comment (gege251) [reply src/MLabsPAB/Files.hs]: Thanks, changed it as suggested.

File: nix/ci.nix (hunk 2 above)
Comment (gege251) [reply nix/ci.nix]: Removed it

**Review by samuelWilliams99 — APPROVED** [review]

**Pull request merged**